### PR TITLE
Fix multiple issues with ddrgen

### PR DESCRIPTION
**Pull request by keithc-ca**
This is in support of https://github.com/eclipse/openj9/issues/378.

Fixes & improvements
* blacklist pattern matching
* close list files
* avoid typedef cycles
* write proper enum size to blob
* handle CR/LF line-ends reading a list files
* pass const references
* favour constructor use over assignment
* formatting
* improve safety
* mark constructors explicit as appropriate
* improve DEBUGPRINTF and ERRMSG macros and use consistent format
* remove redundant or unnecessary #includes
* add constructor for Dwarf_Attribute_s
* add Dwarf_Attribute_s._flag
* add dwarf_formflag() for OSX
* omit static fields from expected superset.out
* improve handling of vtable pointers
* improve 'getmacros' script (and don't let it change timestamps)
* use 'find' to avoid scanning unintersting files
* pull regex definitions out of the inner loop
* fix references to ddr/tools/getmacros
* improve test code
* add an example of map_to_type= without following whitespace
* improve namespace handling
* improve getmacros script
* replace use of 'awk' with 'sed'
* fix @MACRO emissions
* tighten API of ddrgen
* fix validation that macro values are constant
* streamline MacroTool
* fix constant evaluation (including numeric prefixes like '0x')
* fix problem reading macrolist
* use const types where appropriate
* use canonical type expressions
* make as many *.i files as possible
* be careful constructing name to type map for macros
* implicitly blacklist illegal type names
* remove overloading in TypeVisitor
* allow writing of blob and superset independently
* simplify blacklist handling
* make reading blacklist more robust
* fix override support
* fix handling of #undef by getmacros
* ignore linkage_name DWARF attribute
* add ClassType::_isComplete
* construct all fields explicitly
* relocate constants to J9HashtableConstants
* repeat structure for typedef with distinct name

ddrgen
* consider macros in ClassType
* fix Type::renameFieldsAndMacros
* don't require values for enum literals
* don't trim trailing *[] from override value
* generalize support for opaque types
* include typedef with distinct name
* new strategy for eliminating duplicate types
* remove special handling for CLimits
* remove unused _offset field in Modifiers
* reset modifiers with field type override

Blob generation
* don't duplicate fields in blob
* extract common code
* propagate failure to parent process

getmacros
* fix handling of ddr_options: valuesandbuildflags
* improve include guard detection
* fix constant handling

would previously map to the following
(independent of whether A was defined or not)
```
@MACRO_A_DEFINED A_DEFINED 1
@MACRO_A_DEFINED A_DEFINED 0
```

### CODE REVIEW

**Comment by charliegracie** [comment]
@genie-omr build zos zlinux plinux

File: ddr/lib/ddr-ir/Modifiers.cpp [hunk 1]
```diff
@@ -23,54 +23,56 @@
...
-		s += "&";
+	for (size_t i = 0; i < _referenceCount; ++i) {
+		s << "&";
 	}
 
-	if (isArray()) {
-		for (size_t i = 0; i < getArrayDimensions(); i++) {
-			s += "[]";
-		}
+	for (size_t i = getArrayDimensions(); i != 0; i -= 1) {
```
Comment (charliegracie) [inline ddr/lib/ddr-ir/Modifiers.cpp]: I guess getArrayDimensions() returns 0 if isArray() returns false?

File: ddr/lib/ddr-scanner/Scanner.cpp [hunk 2]
```diff
@@ -19,93 +19,131 @@
...
+
+	return false;
+}
+
+static bool
+stringMatchesAny(const string &candidate, const set<string> &patterns)
+{
+	bool match = false; // patterns.find(candidate) != patterns.end();
+
+	if (!match) {
```
Comment (charliegracie) [inline ddr/lib/ddr-scanner/Scanner.cpp]: Can we get rid of match?

File: ddr/lib/ddr-ir/Modifiers.cpp (hunk 1 above)
Comment (keithc-ca) [reply ddr/lib/ddr-ir/Modifiers.cpp]: Yes.

File: ddr/lib/ddr-scanner/Scanner.cpp (hunk 2 above)
Comment (keithc-ca) [reply ddr/lib/ddr-scanner/Scanner.cpp]: Will do.

**Comment by keithc-ca** [comment]
@charliegracie Any more feedback or questions?

**Pull request merged**